enhancements/cvo/object-removal-manifest-annotation.md

```diff
@@ -0,0 +1,134 @@
+
+```
+apiVersion: apps/v1
+...
+metadata:
+...
+  annotations:
+    release.openshift.io/delete: "true"
+```
+
+### User Stories
+
+#### Story 1
+
+Remove the cluster-autoscaler-operator deployment.
+The existing cluster-autoscaler-operator deployment manifest 0000_50_cluster-autoscaler-operator_07_deployment.yaml is modified to contain the delete annotation:
+```
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: cluster-autoscaler-operator
+  namespace: openshift-machine-api
+  annotations:
+    release.openshift.io/delete: "true"
+...
+```
+Additional manifest properties such as `spec` may be set if convenient (e.g. because you are looking to make a minimal change vs. a previous version of the manifest), but those properties have no affect on manifests with the delete annotation.
+
+#### Story 2
+
+In release 4.5 two jobs, openshift-service-catalog-controller-manager-remover and openshift-service-catalog-apiserver-remover, were created to remove the Service Catalog.
+For release 4.6 these jobs and all their supporting cluster objects should be removed.
+This enhancement allows a user to accomplish this in a very simple manner by modifying all the manifests associated with these two jobs to contain the delete annotation.
+In this case the manifests to be modified are:
+* 0000_50_cluster-svcat-apiserver-operator_04_roles.yaml
+* 0000_50_cluster-svcat-apiserver-operator_05_serviceaccount.yaml
+* 0000_50_cluster-svcat-controller-manager-operator_06_roles.yaml
+* 0000_50_cluster-svcat-controller-manager-operator_07_serviceaccount.yaml
+* 0000_90_cluster-svcat-apiserver-operator_01_remover_job.yaml
+* 0000_90_cluster-svcat-controller-manager-operator_01_remover_job.yaml
+
+### Implementation Details/Notes/Constraints
+
+CVO will use its existing logic to discover and identify given manifests and to handle any associated errors.
+Once the manifest has been properly identified CVO will check for the delete annotation release.openshift.io/delete.
+If found and set to "true" the associated object will be removed from the cluster.
+All other annotation values will result in a CVO error and no further processing of the manifest.
+
+### Risks and Mitigations
+
+The only risk is human error resulting in an incorrect annotation being placed in a manifest.
+This results in a CVO error and no further processing of the manifest.
+
+## Design Details
+
+### Test Plan
+
+To be done
+
+### Graduation Criteria
+
+GA. When it works, we ship it.
+
+### Upgrade / Downgrade Strategy
+
+No special handling is needed as long as no change occurs to Kubernetes annotations.
+
+### Version Skew Strategy
+
+No special consideration.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+To be done
```

Comment:
I'd like to see an alternative considered that provides a manifest with a different type that references a list of resources by group, resource, name and provides a deletion ordering guarantee.

Reply:
Use case?  :)

Reply:
Why does deletion need to guarantee ordering when creation does not. The user requesting the deletes should just use the existing CVO ordering scheme https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#what-is-the-order-that-resources-get-createdupdated-in  and, assuming they're deleting everything originally created, just reverse the order.
This should minimize code changes to CVO as well.

Reply:
Reverse ordering is required as david notes because you have to work inside out.  Because of that, you can't run it in the normal flow safely, although it's trivial to reorder the graph backwards with deletes (not focused on the how yet, but the what and why)

Reply:
Go that reverse order is required. My comment is focused on "...and provides a deletion ordering guarantee". I'm simply saying we will reuse the current CVO ordering scheme, not create something new. This of course puts the onus on the dev to name their delete manifests correctly but that's no different than the responsibility one has when they originally created the manifests. So given you adhere to the scheme CVO will delete in the order you specify as it does for creation.